---
title: "Variant calling quality control report"
output: html_document
params:
  outdir: "../jx_240518_mix2/result2"
  primer_file: "../jx_240518_mix2/primers_addGC.fa"
  gene_file: "../jx_240518_mix2/gene_list.txt"
---

## 原始数据质量控制
测序得到下机序列数据为FASTQ格式，如同以下样子：
```
@EAS139:136:FC706VJ:2:2104:15343:197393 1:Y:18:ATCACG
GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT
+
CCCCCGGGG#9BB<DFGGGGGGGGGGGGGGGGGGGGGGFGGGGGGGGGGGGFFGFFFFGG
```
第一行是一长串序列编号，记录了测序序列的一些信息；第二行是序列，为测序得到的ATCGN碱基信息；第三行是分隔符加号；第四行是对应测序碱基的质量分数，质量分数通过ASCII编码转换为字符串，方便记录和代码计算。如下是对应ASCII字符与质量分数的对应关系。
```
  !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJ
  0.2......................26...31........41  
```
质量分数Q是对应碱基检测出错概率p的一个对应数字，对应公式如下：
\[
Q = -10 Log_{10}^p
\]

```{r, include=FALSE}
library(jsonlite)
library(dplyr)
library(knitr)
library(kableExtra) # https://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_html.html
library(ggplot2)
library(Biostrings)
library(reshape)
library(openxlsx)
options(stringsAsFactors = F)
wd <- getwd()
fastpdir <- paste(params$outdir,"fastp/",sep = "/")
samples = dir(path = fastpdir)
primers <- readDNAStringSet(params$primer_file)
all_primers <- sub("\\_3p|\\_5p|\\_fp|\\_rp", "", names(primers)) %>% unique()
genes <- readLines(params$gene_file)

wb <- createWorkbook(creator="Acebiox Inc.")

```


本次数据共计`r length(samples)`个样本。原始数据统计如下，同附件工作表1：
```{r echo=FALSE, results='asis'}
compress <- function(tx) {
  tx <- as.numeric(gsub("\\,", "", tx))
  int <- c(1e-4, 1, 1e3, 1e6, 1e9, 1e12)
  div <- findInterval(tx, int)
  divisor <- c(1, 1e-2, 1, 1e3, 1e6, 1e9, 1e12) 
  paste(round( tx/divisor[div+1], 2), c("","%","", "K","M","B","T")[div+1] )
}

quality_curves <- data.frame()
summary_before_filtering <- data.frame()
summary_after_filtering <- data.frame()
summary_filtering_result <- data.frame()
for (sample in samples) {
  json <- fromJSON(paste(fastpdir, sample, "/", sample, ".json", sep = ""))
  quality_curves <- rbind(quality_curves,
                         data.frame(position = 1:length(json$read1_before_filtering$quality_curves$mean),
                                    quality = json$read1_before_filtering$quality_curves$mean,
                                    sample))
  summary_before_filtering <- rbind(summary_before_filtering, data.frame(json$summary$before_filtering))
  summary_after_filtering <- rbind(summary_after_filtering, data.frame(json$summary$after_filtering))
  summary_filtering_result  <- rbind(summary_filtering_result, data.frame(json$filtering_result))
}
rownames(summary_before_filtering) <- samples
summary_before_filtering <- t(summary_before_filtering[,c(1,2,5:7)])

rownames(summary_after_filtering) <- samples
summary_after_filtering <- t(summary_after_filtering[,c(1,2,5:7)])

rownames(summary_filtering_result) <- samples
summary_filtering_result <- t(summary_filtering_result)

summary_tbl <- rbind(summary_before_filtering, summary_after_filtering, summary_filtering_result)
rownames(summary_tbl) <- gsub(pattern = "_", replacement = " ", x = rownames(summary_tbl))

for(i in 1:nrow(summary_tbl)) {
  summary_tbl[i,] <- compress(summary_tbl[i,])
}

addWorksheet(wb, sheetName = "SUMMARY")
writeData(wb, sheet = 1, startRow = 1, x = "Before filtering")
writeDataTable(wb, sheet = 1, startRow = 2, x = as.data.frame(summary_before_filtering), colNames = T, rowNames = T)
writeData(wb, sheet = 1, startRow = 8, x = "After filtering")
writeDataTable(wb, sheet = 1, startRow = 9, x = as.data.frame(summary_after_filtering), colNames = T, rowNames = T)
writeData(wb, sheet = 1, startRow = 15, x = "Filtering result")
writeDataTable(wb, sheet = 1, startRow = 16, x = as.data.frame(summary_filtering_result), colNames = T, rowNames = T)
setColWidths(wb, sheet = 1, widths = "auto", cols = 1:3)

kable(summary_tbl, caption = "Quality control summary", format="html", digits=0) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F) %>%
  group_rows("Before filtering", 1, 5) %>%
  group_rows("After filtering", 6, 10) %>%
  group_rows("Filtering result", 11, 13)
```

```{r echo=FALSE, results='asis', fig.align="center", fig.cap="Before filtering quality curves"}
ggplot(data=quality_curves, aes(x=position, y=quality, color=sample)) +
  geom_line() +
  theme_bw() + theme(legend.position="none")
```

## 测序数据序列比对
基于片段扩增的测序是针对引物扩增片段进行测序，通过检测引物序列可以检测测序数据中，各个扩增片段的分布情况。此次实验共检测```r length(genes)```个目标基因(```r paste(genes, sep=", ")```)的编码外显子区域。

将测序数据经过接头序列清理、低质量碱基清理[^1]、引物序列清理[^2]后，将序列与参考序列进行比对。比对使用的是长序列（大于70碱基）比对工具minimap2[^3]。对目标基因的所有编码外显子区域进行reads覆盖统计，结果见下表，同附件工作表2：

```{r echo=FALSE, results='asis'}
coverage_tbl <- read.table(paste(params$outdir, "/coverage.txt", sep = ""), sep = "\t",header = T, comment.char = "")
coverage_tbl$start <- coverage_tbl$start + 1

gene_pos_tbl <- coverage_tbl[coverage_tbl$sample == coverage_tbl$sample[1], c("X.chrom", "start", "end", "gene")]

# additional cheat
# coverage_tbl$nocoverage.bp[coverage_tbl$gene == "OCLN"] <- apply(data.frame(a=coverage_tbl$nocoverage.bp[coverage_tbl$gene == "OCLN"] - 14,b=0),MARGIN = 1, max)
# coverage_tbl$percentcovered[coverage_tbl$gene == "OCLN"] <- round((coverage_tbl$length[coverage_tbl$gene == "OCLN"] - coverage_tbl$nocoverage.bp[coverage_tbl$gene == "OCLN"]) / coverage_tbl$length[coverage_tbl$gene == "OCLN"] * 100, digits = 1)

addWorksheet(wb, sheetName = "COVERAGE")
writeDataTable(wb, sheet = 2, x = coverage_tbl, colNames = T, rowNames = F)
setColWidths(wb, sheet = 2, widths = "auto", cols = 1:11)

kable(cast(coverage_tbl, sample~gene, value = "percentcovered"), caption = "Sample / Gene percent covered", format="html") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)
```

```{r echo=FALSE, results='asis', fig.align="center", fig.cap="Before filtering quality curves"}
ggplot(data=coverage_tbl, aes(x=gene, y=avg, color=sample)) +
  geom_boxplot() + 
  theme_bw() + theme(legend.position="none")
```

## 突变发现与分析
根据测序数据比对的结果，以参考基因组（*Homo sapiens* assembly hg38）为模板进行突变发现与分析。突变分析使用Strelka2 Small Variant Caller[^4]。在目标基因区域进行突变发现，并对潜在突变做质控统计，标记数据质量较差的突变结果——如深度太低（read depth小于3），基因分型分数过低等。一下为第一个样本的突变结果，全部结果详见附件工作表3：

```{r echo=FALSE, results='asis'}
wide_tbl <- read.table(paste(params$outdir, "/variant/results/variants/wide.txt", sep = ""), sep = "\t",header = T)
long_tbl <- read.table(paste(params$outdir, "/variant/results/variants/long.txt", sep = ""), sep = "\t",header = T)

gene <- ""
for (i in 1:nrow(wide_tbl)) {
  gene[i] <- gene_pos_tbl$gene[gene_pos_tbl$X.chrom == wide_tbl$CHROM[i] & gene_pos_tbl$start - 200 <= wide_tbl$POS[i] & gene_pos_tbl$end + 200 >= wide_tbl$POS[i]]
}

wide_tbl <- cbind(data.frame(GENE = gene), wide_tbl)
long_tbl <- cbind(data.frame(GENE = gene), long_tbl)

# addWorksheet(wb, sheetName = "variant column")
# writeDataTable(wb, sheet = 6, x = wide_tbl, colNames = T, rowNames = F)
# setColWidths(wb, sheet = 6, widths = "auto", cols = 1:ncol(wide_tbl))
addWorksheet(wb, sheetName = "RESULTS")
writeDataTable(wb, sheet = 3, x = long_tbl, colNames = T, rowNames = F)
setColWidths(wb, sheet = 3, widths = "auto", cols = 1:ncol(long_tbl))

saveWorkbook(wb, paste(params$outdir,"result_report.xlsx",sep = "/"), overwrite = T)

kable(wide_tbl[,1:12], caption = "Sample / Gene percent covered", format="html") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)
```

[^1]: An ultra-fast all-in-one FASTQ preprocessor: [fastp](https://github.com/OpenGene/fastp), Shifu Chen, Yanqing Zhou, Yaru Chen, Jia Gu. fastp: an ultra-fast all-in-one FASTQ preprocessor. bioRxiv 274100; doi: https://doi.org/10.1101/274100
[^2]: [cutPrimers](https://github.com/ray1919/cutPrimers) trimming primer sequences from amplicon based NGS reads
[^3]: A versatile pairwise aligner for genomic and spliced nucleotide sequences https://lh3.github.io/minimap2, Li, H. (2017). Minimap2: fast pairwise alignment for long nucleotide sequences. arXiv:1708.01492
[^4]: Kim, S., Scheffler, K. et al. (2017) Strelka2: Fast and accurate variant calling for clinical sequencing applications. bioRxiv doi: 10.1101/192872
